### PR TITLE
feat(#966): Adds JUnit 5 support for Container Object Pattern DSL

### LINKED ISSUE #966: Create JUnit5 support for ContainerDsl
##### Issue Overview

Create JUnit5 support for ContainerDsl

##### Expected Behaviour

Be able to use JUnit5 with Cube Docker Container Dsl

##### Current Behaviour

It can only be used with JUnit4 `@Rule`

### PR DESCRIPTION
**Pull request by lordofthejars**
#### Short description of what this resolves:

Adds JUnit 5 support for Container Object Pattern DSL

#### Changes proposed in this pull request:

- Adds JUnit 5 module creating an extension for Container DSL and Network DSL
- Create `ftest` showing the features

Fixes #966

### CODE REVIEW

File: docker/junit-rule/src/main/java/org/arquillian/cube/docker/junit/Reflections.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+public class Reflections {
+
+    public static final void injectObject(Object instance, Field field, Object injection) throws IllegalAccessException {
```
Comment (dipak-pawar) [inline docker/junit-rule/src/main/java/org/arquillian/cube/docker/junit/Reflections.java]: static method can't be overridden, as they are associated with class. Do we have any reason to make it final

File: docker/junit-rule/src/main/java/org/arquillian/cube/docker/junit/Reflections.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+        field.setAccessible(true);
+        field.set(instance, injection);
+
+    }
+
+    public static final List<Field> findAllFieldsOfType(Class<?> clazz, Class<?> fieldType, Predicate<Field> restriction) {
```
Comment (dipak-pawar) [inline docker/junit-rule/src/main/java/org/arquillian/cube/docker/junit/Reflections.java]: same as above

File: docker/junit-rule/src/main/java/org/arquillian/cube/docker/junit/Reflections.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        while (currentClass != Object.class) {
+            currentClass = currentClass.getSuperclass();
+            fields.addAll(findAllFieldsFilteringByType(currentClass, fieldType, restriction));
+        }
+
+        return fields;
+
+    }
+
+    private static final List<Field> findAllFieldsFilteringByType(Class<?> clazz, Class<?> fieldType, Predicate<Field> restriction) {
```
Comment (dipak-pawar) [inline docker/junit-rule/src/main/java/org/arquillian/cube/docker/junit/Reflections.java]: same as above

File: docker/junit-rule/src/main/java/org/arquillian/cube/docker/junit/Reflections.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        return Arrays.stream(clazz.getDeclaredFields())
+            .peek(f -> f.setAccessible(true))
+            .filter(f -> f.getType().equals(fieldType))
+            .filter(restriction)
+            .collect(Collectors.toList());
+    }
+
+    public static final Optional<Field> findFieldByGenericType(Class<?> clazz, Class<?> fieldType, Class<?> genericType) {
```
Comment (dipak-pawar) [inline docker/junit-rule/src/main/java/org/arquillian/cube/docker/junit/Reflections.java]: same as above

File: docker/junit5/src/main/java/org/arquillian/cube/docker/junit5/ContainerDsl.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+import org.arquillian.cube.docker.impl.client.config.Await;
+import org.arquillian.cube.docker.impl.client.containerobject.dsl.BindMode;
+import org.arquillian.cube.docker.impl.client.containerobject.dsl.Container;
+import org.arquillian.cube.docker.impl.client.containerobject.dsl.ContainerBuilder;
+import org.arquillian.cube.docker.junit.rule.ContainerDslRule;
+import org.arquillian.cube.docker.junit.rule.NetworkDslRule;
+import org.arquillian.cube.spi.CubeOutput;
+import org.jboss.shrinkwrap.api.Archive;
+
+public class ContainerDsl {
```
Comment (dipak-pawar) [inline docker/junit5/src/main/java/org/arquillian/cube/docker/junit5/ContainerDsl.java]: How about writing tests for `ContainerDsl`?

File: docker/junit5/src/main/java/org/arquillian/cube/docker/junit5/NetworkDsl.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
+package org.arquillian.cube.docker.junit5;
+
+import org.arquillian.cube.docker.impl.client.containerobject.dsl.Network;
+import org.arquillian.cube.docker.impl.client.containerobject.dsl.NetworkBuilder;
+
+public class NetworkDsl {
```
Comment (dipak-pawar) [inline docker/junit5/src/main/java/org/arquillian/cube/docker/junit5/NetworkDsl.java]: How about writing tests for `NetworkDsl`?

File: docker/junit-rule/src/main/java/org/arquillian/cube/docker/junit/Reflections.java (hunk 1 above)
Comment (lordofthejars) [reply docker/junit-rule/src/main/java/org/arquillian/cube/docker/junit/Reflections.java]: https://stackoverflow.com/questions/1743715/behaviour-of-final-static-method

File: docker/junit-rule/src/main/java/org/arquillian/cube/docker/junit/Reflections.java (hunk 2 above)
Comment (lordofthejars) [reply docker/junit-rule/src/main/java/org/arquillian/cube/docker/junit/Reflections.java]: https://stackoverflow.com/questions/1743715/behaviour-of-final-static-method

File: docker/junit-rule/src/main/java/org/arquillian/cube/docker/junit/Reflections.java (hunk 3 above)
Comment (lordofthejars) [reply docker/junit-rule/src/main/java/org/arquillian/cube/docker/junit/Reflections.java]: https://stackoverflow.com/questions/1743715/behaviour-of-final-static-method

File: docker/junit-rule/src/main/java/org/arquillian/cube/docker/junit/Reflections.java (hunk 4 above)
Comment (lordofthejars) [reply docker/junit-rule/src/main/java/org/arquillian/cube/docker/junit/Reflections.java]: https://stackoverflow.com/questions/1743715/behaviour-of-final-static-method

File: docker/junit5/src/main/java/org/arquillian/cube/docker/junit5/ContainerDsl.java (hunk 5 above)
Comment (lordofthejars) [reply docker/junit5/src/main/java/org/arquillian/cube/docker/junit5/ContainerDsl.java]: I thought the same but they are really POJO classes and I don't see much benefit on writing tests for these classes when we have a functional test.

File: docker/junit5/src/main/java/org/arquillian/cube/docker/junit5/NetworkDsl.java (hunk 6 above)
Comment (lordofthejars) [reply docker/junit5/src/main/java/org/arquillian/cube/docker/junit5/NetworkDsl.java]: Same

Comment (dipak-pawar) [reply docker/junit5/src/main/java/org/arquillian/cube/docker/junit5/NetworkDsl.java]: this is fine for `NetworkDsl`

File: docker/junit5/src/main/java/org/arquillian/cube/docker/junit5/ContainerDsl.java (hunk 5 above)
Comment (dipak-pawar) [reply docker/junit5/src/main/java/org/arquillian/cube/docker/junit5/ContainerDsl.java]: If we write some tests around `ContainerDsl`, we'll tests POJO and if any logic around it. Also it'll helpful for end user, if he check the tests around this, end user will get clear idea about how can he use it?

Comment (lordofthejars) [reply docker/junit5/src/main/java/org/arquillian/cube/docker/junit5/ContainerDsl.java]: Ok I'll write some tests for this case.

File: docker/junit-rule/src/main/java/org/arquillian/cube/docker/junit/Reflections.java (hunk 1 above)
Comment (dipak-pawar) [reply docker/junit-rule/src/main/java/org/arquillian/cube/docker/junit/Reflections.java]: @lordofthejars What's the purpose here to make static method final?

Comment (lordofthejars) [reply docker/junit-rule/src/main/java/org/arquillian/cube/docker/junit/Reflections.java]: The keyword final will disable the method from being hidden. So they cannot be hidden and an attempt to do so will result in a compiler error. `final` modifier is just a compilation time modifier it is dropped in code generation phase, so it is just a helper operator for developers. In this case if anybody tries to extend reflections class and `create` a method with same name, compiler will warn (error) saying that you are doing something that might confuse other people when read the code.

Comment (dipak-pawar) [reply docker/junit-rule/src/main/java/org/arquillian/cube/docker/junit/Reflections.java]: thx for explanation. Now this is clear to me

**Pull request merged**